### ISSUE TITLE
Migrate to WebExtensions, if it's possible.

### ISSUE THREAD

**Issue by AlexBeLi** [lazierthanthou/sqlite-manager#75]
If not, create notification on addons.mozilla.org what addon will stop working in Firefox 57 with the Webextensions exclusive requirement for addons.

**Comment by Pomax** [comment]
It's been quite a few months: @lazierthanthou had you see this issue? If so, is this planned or is contributor help required for you to get this work done before Firefox turns off the old API in a few weeks time?

**Comment by justinclift** [comment]
Well... this project seems pretty unmaintained.  There have been a few PR's accepted a while ago, but nothing substantial.

If this project doesn't get rewritten in time to work with FF 57, the SQLite GUI project I'm putting time into is actively maintained:

&nbsp; &nbsp; https://github.com/sqlitebrowser/sqlitebrowser

It's not browser based though, so might not be suitable for people really wanting that (unsure).

**Comment by lazierthanthou** [comment]
FF 57 does not support access to sqlite through extensions. Hence, the present state of the project.

**Comment by justinclift** [comment]
Ahhh, that explains things.  Wonder how many other FireFox extensions will be outright killed due to the change?  Sounds like it'll be painful for a lot of people. 😦 

When FireFox 57 rolls out, people that need a supported browser but don't want to lose access to SQLite Manager can probably use FireFox ESR (Extended Support Release):

&nbsp; &nbsp; https://www.mozilla.org/en-US/firefox/organizations/all/

It's a version of Firefox normally used by larger groups of users with their own IT, who need something with a slower release pace.  It will be a few more more months until that reaches version 57, which would give people extra time to figure out an alternative.

**Comment by justinclift** [comment]
@lazierthanthou As a thought, it might be useful to put some kind of clear notice on the front page of the project repo, and/or even the Firefox extension page, so people know what to expect and aren't left wondering.

**Comment by Pomax** [comment]
@lazierthanthou what do you mean with "[it] does not support access to sqlite through extensions"? Do you mean that it doesn't allow file access so that sqlite database files can be loaded, or that it does not allow you to run the sqlite driver itself for unpacking those files and visualising the data in the SQLite Manager UI?

**Comment by lazierthanthou** [comment]
FF uses sqlite but the equivalent of mozIStorageService and Storage API (https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Storage) does not provide for sql queries, Please refer to https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Comparison_with_XUL_XPCOM_extensions.

**Comment by lazierthanthou** [comment]
There is a NativeMessaging facility (https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Native_messaging) which would require separate installation of sqlite, but in that case too incoming messages are limited to 1MB which may be inadequate for a number of databases containing sizeable blobs.

**Comment by Pomax** [comment]
Would it make sense to file a bugzilla issue over this, given that SQLite Manager is pretty much the most functional sqlite database inspector and maintenance tool available to a web dev? (it one has not already been filed)

**Comment by justinclift** [comment]
> ... the most functional sqlite database inspector and maintenance tool available to a web dev?

Is that because it's inside Firefox?  Note, not challenging that.  Just trying to understand, so if there's some aspect of things we (sqlitebrowser.org) can do better, we get to understand it and think it through. :smile:

**Comment by Pomax** [comment]
Both that, and the fact that it's a really good and super elaborate UI (with its own large window, sqlite manager has the same feel as a mysqladmin, pgadmin3, etc, rather than "a small utility") that also just "looks like any other program" rather than a hand-spun UI that doesn't really look like a windows app on windows, nor like a mac app on osx. That makes a huge difference in experience: it feels like a normal, straight up full blown administration application.

**Comment by Croydon** [comment]
Maybe https://github.com/kripken/sql.js/ can help?

**Comment by justinclift** [comment]
@Pomax No worries. :smile:

**Comment by Croydon** [comment]
@lazierthanthou Please check out this online demo https://kripken.github.io/sql.js/GUI/

I'm pretty sure that you could do this with the existing APIs, reading file via "upload", keep it in cache, after manipulations write it via the download api. https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/downloads/download

Of course you need to rewrite the UI via HTMl/CSS/JS as well, this might be quite some work but it should be doable to port the add-on to a WebExtension today.

**Comment by wbt** [comment]
Another reason why loss of this extension creates loss of functionality not replaceable by @justinclift 's  sqlitebrowser is that it didn't require admin privileges to install.  

As noted [here](http://forums.mozillazine.org/viewtopic.php?p=14775423#p14775423), there's a [meeting ](https://wiki.mozilla.org/Add-ons/Contribute/Triage) starting in a [few hours](https://docs.google.com/document/d/1g3RMfKZ3671NcusMqkoOiKwfPekRe-VI7Rzqxo6F_Ao/edit#heading=h.hhpni8ijl0wx) for review of what additional WebExtensions APIs might be needed.

**Comment by justinclift** [comment]
@wbt Well... technically with our program it's possible to just extract the contents of the installer and throw that into a directory somewhere.  It'll work fine.  The admin perms are really only needed on Windows, and only for the installer itself.  On that note someone requested a .zip file version of the distribution, but it didn't seem to be much in demand so we never got around to it.  If that would actually be well used, we can probably make a .zip version.  It shouldn't be hard. :smile:

The WebExtensions meeting approach though might work too.  Give it a go. :smile:

**Comment by monk-time** [comment]
@justinclift You should add your app on portablefreeware.com then. :)

**Comment by justinclift** [comment]
@monk-time Thanks. :smile:

Hadn't come across portablefreeware.com before.  Sounds somewhat similar to portableapps.com, which we have a download for too.  Haven't really tried the portableapps.com download personally though, so I don't know that much about it.

**Comment by Xtigyro** [comment]
Are we going to see this very useful add-on revived for the latest Firefox? What's the plan?

**Comment by Pomax** [comment]
@Xtigyro presumably unless you know of a way to get around https://github.com/lazierthanthou/sqlite-manager/issues/75#issuecomment-328328216, the answer is going to stay "no".

**Comment by francipvb** [comment]
hello @justinclift 

> Is that because it's inside Firefox? Note, not challenging that. Just trying to understand, so if there's some aspect of things we (sqlitebrowser.org) can do better, we get to understand it and think it through. 😄

I tried your project but I cannot use it because screen reader support is completely broken, at least in the table creation dialog.

I'll post an issue to your repo  regarding this later.

Cheers,

**Comment by justinclift** [comment]
> ... screen reader support is completely broken ...

Yeah, we've not (yet) thought through our screen reader support at all.

> I'll post an issue to your repo regarding this later.

Thanks @francipvb, please do.

We use Qt 5 for our GUI, which from rough memory has fairly lousy support for screen readers.

If it's reasonably possible to get a Qt5 application working decently with screen readers though, we'd likely be up for it.  Baring in mind that like many OSS projects we're volunteer based, so it generally takes time to get major improvements fully worked through.

If you're willing to keep involved with the process (eg testing stuff and providing real world insight as it gets added), that generally helps greatly. :smile:
